fdbserver/RestoreMaster.actor.cpp

```diff
@@ -0,0 +1,1338 @@
+	state int index = 0;
+	loop {
+		try {
+			wait(delay(1.0));
+			index = 0;
+			std::vector<Future<RestoreCommonReply>> cmdReplies;
+			for(auto& loaderInterf : self->loadersInterf) {
+				self->cmdID.nextCmd();
+				printf("[CMD:%s] Node:%s askLoadersToCollectRestoreAppliersInterfaces for node (index=%d uid=%s)\n", 
+						self->cmdID.toString().c_str(), self->describeNode().c_str(),
+						index, loaderInterf.first.toString().c_str());
+				cmdReplies.push_back( loaderInterf.second.collectRestoreRoleInterfaces.getReply(RestoreSimpleRequest(self->cmdID)) );
+				index++;
+			}
+			std::vector<RestoreCommonReply> reps = wait( timeoutError(getAll(cmdReplies), FastRestore_Failure_Timeout) );
+			printf("[setWorkerInterface] Finished\n");
+			break;
+		} catch (Error &e) {
+			fprintf(stdout, "[ERROR] Node:%s, Commands before cmdID:%s error. error code:%d, error message:%s\n", self->describeNode().c_str(),
+					self->cmdID.toString().c_str(), e.code(), e.what());
+			printf("Node:%s waits on replies time out. Current phase: setWorkerInterface, Retry all commands.\n", self->describeNode().c_str());
+		}
+	}
+
+	return Void();
+}
+
+
+
+// TODO: Revise the way to collect the restore request. We may make it into 1 transaction
+ACTOR Future<Standalone<VectorRef<RestoreRequest>>> collectRestoreRequests(Database cx) {
+	state int restoreId = 0;
+	state int checkNum = 0;
+	state Standalone<VectorRef<RestoreRequest>> restoreRequests;
+	state Future<Void> watch4RestoreRequest;
+
+	//wait for the restoreRequestTriggerKey to be set by the client/test workload
+	state ReadYourWritesTransaction tr(cx);
+
+	loop {
+		try {
+			tr.reset();
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			// Assumption: restoreRequestTriggerKey has not been set
+			// Question: What if  restoreRequestTriggerKey has been set? we will stuck here?
+			// Question: Can the following code handle the situation?
+			// Note: restoreRequestTriggerKey may be set before the watch is set or may have a conflict when the client sets the same key
+			// when it happens, will we  stuck at wait on the watch?
+
+			watch4RestoreRequest = tr.watch(restoreRequestTriggerKey);
+			wait(tr.commit());
+			printf("[INFO][Master] Finish setting up watch for restoreRequestTriggerKey\n");
+			break;
+		} catch(Error &e) {
+			printf("[WARNING] Transaction for restore request in watch restoreRequestTriggerKey. Error:%s\n", e.name());
+			wait(tr.onError(e));
+		}
+	};
+
+
+	loop {
+		try {
+			tr.reset();
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			// Assumption: restoreRequestTriggerKey has not been set
+			// Before we wait on the watch, we must make sure the key is not there yet!
+			//printf("[INFO][Master] Make sure restoreRequestTriggerKey does not exist before we wait on the key\n");
+			Optional<Value> triggerKey = wait( tr.get(restoreRequestTriggerKey) );
```

Comment:
If you can get a new read version and read a key, I'd expect that the watch should have already triggered, so I'd think that just waiting on the watch should be the more efficient way to do this?